### PR TITLE
Add a check for available RAM in the builder image

### PR DESCRIPTION
**Pull request by vladdy**
The issue with builds in Docker randomly failing for people due to available RAM is not rare and although we have the requirement stated in the docs cockroachdb/cockroach#40232, it is helpful to have a check for the available RAM size. This change adds the check for "make build" and "make" parameters passed to the builder script. 

Here is an example of the message when the available RAM is less than required:
```
Running make with -j4
GOPATH set to /go
Cleaning old generated files.
GOFLAGS= go build -o cockroach -v  -tags ' gss make x86_64_pc_linux_gnu' -ldflags '-X github.com/cockroachdb/cockroach/pkg/build.typ=development -extldflags "" -X "github.com/cockroachdb/cockroach/pkg/build.tag=v19.2.0-alpha.20190606-1797-gb97bc55-dirty" -X "github.com/cockroachdb/cockroach/pkg/build.rev=b97bc556e4751983f0e1a7abe0774b6e995a8fa1" -X "github.com/cockroachdb/cockroach/pkg/build.cgoTargetTriple=x86_64-pc-linux-gnu"   -X "github.com/cockroachdb/cockroach/pkg/build.utcTime=2019/08/28 14:00:06"' ./pkg/cmd/cockroach
github.com/cockroachdb/cockroach/pkg/sql/exec
go build github.com/cockroachdb/cockroach/pkg/sql/exec: /usr/local/go/pkg/tool/linux_amd64/compile: signal: killed
Makefile:883: recipe for target 'cockroach' failed
make: *** [cockroach] Error 1
Note: if the build seems to terminate for unclear reasons, note that your container limits RAM to 2046844kB. This may be the cause of the failure. Try increasing the limit to 4GB or above.
```

### CODE REVIEW

File: build/builder.sh [hunk 1]
```diff
@@ -164,6 +164,18 @@ gid=$(id -g)
...
 
+# Build container needs to have at least 4GB of RAM available to compile the project
+# successfully, which is not true in some cases (i.e. Docker for MacOS by default).
+# This wraps the passed build command into a check for avilable memory for "make"
+# and "make build"
+ram_message="Note: if the build seems to terminate for unclear reasons, \
+note that your container limits RAM to \${ram}kB. This may be \
+the cause of the failure. Try increasing the limit to 4GB or above."
+
+cmd="ram=\$(awk '/MemTotal/{print \$2}' /proc/meminfo); $* \
```
Comment (knz) [inline build/builder.sh]: `$* ||` -> `\"$@\" ||`

however overall I am not fond of having the check inside the "main" builder invocation. Instead what you can do is...

File: build/builder.sh [hunk 2]
```diff
@@ -178,4 +190,4 @@ docker run --init --privileged -i ${tty-} --rm \
   --env="TZ=America/New_York" \
   --env=COCKROACH_BUILDER_CCACHE \
   --env=COCKROACH_BUILDER_CCACHE_MAXSIZE \
-  "${image}:${version}" "$@"
+  "${image}:${version}" bash -c "$cmd"
```
Comment (knz) [inline build/builder.sh]: keep `\"$@\"` here, and then below:

```sh
res=$?
if test $res != 0; then
   ram=$(docker run --rm "${image}:${version}" awk '/MemTotal/{print $2}' /proc/meminfo)
   if test "x$1" = "xmake" -a $ram -lt 4000000; then
      echo $ram_message >&2   # be mindful of printing to stderr, not stdout
   fi
fi
exit $res
```

**Review by knz — APPROVED** [review]
LGTM but:

1. feel free to put the 2nd invocation of `docker` "under" the `if test $1 = make` so as to avoid paying that price when running another command. 

2. you may want to include `mkrelease` as another command that triggers the check.

**Review by irfansharif — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40291)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @vladdy)

___
*[build/builder.sh, line 196 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/40291#-LnYIck450R1akkz1rxP:-LnYIck5CY54YL9febxt:b-jzo4n8) ([raw file](https://github.com/cockroachdb/cockroach/blob/294cd8f5dee6de55805940710a21c432a7b3eb91/build/builder.sh#L196)):*
> ```Shell
>    ram=$(docker run -i --rm \
>          -u "$uid:$gid" \
>          "${image}:${version}" awk '/MemTotal/{print $2}' /proc/meminfo)
> ```

nit: I believe something like`awk '/MemTotal/{printf "%.2f", $2/1024/1024}' /proc/meminfo` would give you free space in GB.

**Comment by knz** [comment]
xref #40378

**Comment by knz** [comment]
I updated your permission please try again

**Review by vladdy — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40291)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @knz)

___
*[build/builder.sh, line 175 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40291#-LnNxHGRcUDV3h0cPOQ5-r1-175:-LnwUqdQ5uQT_Ux4ovE4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d5ce3494d78f8bcc710e4b6f2c2f3ad0e87435e/build/builder.sh#L175)):*
<details><summary><i>Previously, vladdy (Vlad Artamonov) wrote…</i></summary><blockquote>

👍
</blockquote></details>

Done.
___
*[build/builder.sh, line 193 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40291#-LnNxHGRcUDV3h0cPOQ5-r1-193:-LnwUrlS4oPGEXa2FPqM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d5ce3494d78f8bcc710e4b6f2c2f3ad0e87435e/build/builder.sh#L193)):*
<details><summary><i>Previously, vladdy (Vlad Artamonov) wrote…</i></summary><blockquote>

I'm updating this PR with the proposed changes.
</blockquote></details>

Done.
___
*[build/builder.sh, line 196 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/40291#-LnYIck450R1akkz1rxP:-LnxCKmH6c_PJ3hjovJ8:blb5q8r) ([raw file](https://github.com/cockroachdb/cockroach/blob/294cd8f5dee6de55805940710a21c432a7b3eb91/build/builder.sh#L196)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

nit: I believe something like`awk '/MemTotal/{printf "%.2f", $2/1024/1024}' /proc/meminfo` would give you free space in GB.
</blockquote></details>

Yeah, I thought about that, but "test" does not work with non-integers. However, I'll adjust all unit to be GB in the message.

**Pull request merged**